Added information about CuArrays (#1609)
CliMA/Oceananigans.jl: docs/src/simulation_tips.md

the compiler can't translate them into GPU code and they fail for GPU runs. (This limitation is discussed 
in [this Github issue](https://github.com/CliMA/Oceananigans.jl/issues/1241) and contributors are welcome.)
For example, in the example below, calculating `u²` works in both CPUs and GPUs, but calculating 

> `KE` may not work in some GPUs:

Reviewer: Hmm, I think it works now --- at least, we test it.

Reviewer: Maybe I should put it this way: as far as I know, it's always possible to calculate `KE` on the GPU. If we know of scenarios where it doesn't work, let's raise an issue!

Reviewer: There's a bunch of tests here that all pass:

https://github.com/CliMA/Oceananigans.jl/blob/292da65c92d8b76f03cc95d88b5eb067e41e62a8/test/test_abstract_operations_computed_field.jl#L694-L733

Author: It does work on mine. But I have perceived some difference between GPUs. There are some operations that you report as not working for you that work for me and vice versa. So I think it may not work on all GPUs.

However, if you can think of an interesting example that won't compile on GPUs for me to substitute let me know and I'll change it! I honestly couldn't find any. Everything I've tried that was reasonable compiled successfully (kudos to you for significantly improving compilation!).

Reviewer: I think there are examples on #1241 that are very long ?

Reviewer: How about

```julia
julia> compute!(ComputedField(∂x(u)^2 + ∂y(v)^2 + ∂z(w)^2 + ∂x(w)^2))
ERROR: CUDA error: device kernel image is invalid (code 200, ERROR_INVALID_IMAGE)
```

Reviewer: > So I think it may not work on all GPUs.

I remember you mentioning this but I think we need to carefully test that this is indeed the case before insinuating such in documentation --- because it does not seem possible except maybe for a machine-dependent bug in some low-level library.

Author: > I remember you mentioning this but I think we need to carefully test that this is indeed the case before insinuating such in documentation

I don't know what the state of affairs is right now after your improvements, but when I tested them before I was using the main branch (right after you posted your scripts) and I literally copy-pasted your scripts. So I don't know what room for error there is. But yes, I think we could at some point make a more rigorous test to see if everything is reproducible across all many GPUs as we can get.

Author: Switched the example to use energy dissipation rate. I think it's a better illustration and it'll definitely fails on GPUs.

├── grid: RegularRectilinearGrid{Float64, Periodic, Periodic, Bounded}(Nx=1, Ny=1, Nz=1)
├── tracers: (:T, :S)
├── closure: IsotropicDiffusivity{Float64,NamedTuple{(:T, :S),Tuple{Float64,Float64}}}
├── buoyancy: SeawaterBuoyancy{Float64,LinearEquationOfState{Float64},Nothing,Nothing}
└── coriolis: Nothing

julia> typeof(model.velocities.u.data)
OffsetArrays.OffsetArray{Float64,3,CUDA.CuArray{Float64,3}}

julia> adapt(Array, model.velocities.u.data)
3×3×3 OffsetArray(::Array{Float64,3}, 0:2, 0:2, 0:2) with eltype Float64 with indices 0:2×0:2×0:2:
[:, :, 0] =
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

[:, :, 1] =
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

[:, :, 2] =
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0
```

Notice that in order to view the `CuArray` that stores values for `u` we needed to transform
it into a regular `Array` first using `Adapt.adapt`. If we naively try to view the `CuArray`
without that step we get an error:

```julia
julia> model.velocities.u.data
3×3×3 OffsetArray(::CUDA.CuArray{Float64,3}, 0:2, 0:2, 0:2) with eltype Float64 with indices 0:2×0:2×0:2:
[:, :, 0] =
Error showing value of type OffsetArrays.OffsetArray{Float64,3,CUDA.CuArray{Float64,3}}:

> ERROR: scalar getindex is disallowed

Reviewer: This error can also be alleviated by writing `CUDA.allowscalar(true)`. Scalar operations are legal, but _disabled_ by default because they are slow. Sometimes its nice to enable scalar operations when working at the REPL and prototyping.

Author: Included a remark!

Author: Although I fear this passage is about to become obsolete very soon. Testing these with CUDA.jl 2.6 in Julia 1.6 works out of the box. Even if I explicitly ask CUDA to not allow scalar operations:

```julia
julia> CUDA.allowscalar(false)

julia> A = rand(1:10, (5,5,5));

julia> d_A = CuArray(A);

julia> d_A
5×5×5 CuArray{Int64, 3}:
[:, :, 1] =
  5  4  1  5   5
  9  1  1  1   5
 10  4  6  7  10
 10  5  7  2   3
  8  5  1  8   6

...
```

Reviewer: Yeah, I think it makes a ton of sense to enable scalar operations for printing purposes...

Reviewer: I think the error users get is something like "Could not compile <kernel with long unintelliglbe name> because on of the arguments is not isbits" or something like that. Should we help users understand that this error often occurs because they have tried to use an `Array` on the GPU when they should have used `CuArray`?

Reviewer: Up to you, the changes so far are an improvement as far as I am concerned. This is also not a very common error, because functions are a better, faster choice for implementing boundary conditions and forcing functions. The few users that need to use arrays (mostly for boundary conditions) are doing something sophisticated, like coupling oceananigans to an external model.

ERROR: scalar getindex is disallowed
```

> Here Julia is telling us that using scalar `getindex` is not `allowed`. So another way to go
> around this limitation is to simply allow scalar operations:

Reviewer: ```suggestion
Here CUDA throws an error because scalar `getindex` is not `allowed`. Another way
around this limitation is to allow scalar operations:
```

```julia
julia> using CUDA; CUDA.allowscalar(true)

julia> model.velocities.u.data
3×3×3 OffsetArray(::CuArray{Float64,3}, 0:2, 0:2, 0:2) with eltype Float64 with indices 0:2×0:2×0:2:
[:, :, 0] =
┌ Warning: Performing scalar operations on GPU arrays: This is very slow, consider disallowing these operations with `allowscalar(false)`
└ @ GPUArrays ~/.julia/packages/GPUArrays/WV76E/src/host/indexing.jl:43
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

[:, :, 1] =
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0

[:, :, 2] =
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0
```

Notice the warning we get when we do this. Scalar operations on GPUs can be very slow, so it is
advised to only use this last method when using the REPL or prototyping --- never in
production-ready scripts.

You might also need to keep these differences in mind when using arrays

> to `set!` initial conditions or when using arrays to provide boundary conditions and

Reviewer: We have code that seems aimed at supporting `set!` between a field on the GPU and a field on the CPU:

https://github.com/CliMA/Oceananigans.jl/blob/292da65c92d8b76f03cc95d88b5eb067e41e62a8/src/Fields/set!.jl#L55-L60

So I think the main gotcha (right now, until hopefully we fix it) is using the wrong type of array in boundary conditions, forcing functions, or other types that are used internally by the model during time-stepping.

Author: @ali-ramadhan Wrote that part, so I'm reluctant to change it (since I don't know much about that topic)

Reviewer: I guess you can't `set!` a field on the CPU to a `CuArray`. Is this what's meant? Wrong information in the docs is pretty much the epitome of evil.

Author: I just simplified it. If I understood correctly that's a passage about setting ICs, BCs, etc with an array instead of a function.
